### DISCUSSION TITLE
COP calculation with primary/secondary current

### DISCUSSION THREAD

**Discussion by dandjo** [raomin/ESPAltherma#184] [General]
Hi Altherma Fans.

So far I have calculated the power of the compressor and the COP using "INV Primary Current" and the voltage of my grid. However, this results in large fluctuations and apparently time-shifted data. If I divide the current of the pulsed 3-phase output signal "INV Secondary Current" by 3 and multiply it by the voltage "Voltage (N-phase)", I get much more plausible values for the power of the compressor and ultimately for the COP. Additionally I consider the power factor at lower currents. I packed this data into a function based on some measurement data from my smart meter. Here is the result of the compressor power calculation (openHAB ECMAScript).

```
var current = parseFloat(itemRegistry.getItem('espaltherma_compressor_secondary_current').state);
var voltage = parseFloat(itemRegistry.getItem('espaltherma_compressor_voltage_n_phase').state);
var emergency = itemRegistry.getItem('espaltherma_emergency_active').state;

current = current / 3;  // 3-phase sum

var power = 0;

if (current > 0 && voltage > 0 && emergency === OFF) {
  var power_factor = 0.752385 * Math.pow(current, 0.11962) - 0.00141164;
  power_factor = power_factor > 1 ? 1 : power_factor;
  
  power = current * voltage * power_factor;
  power = Math.round(power * 100) / 100;
}

events.sendCommand('espaltherma_compressor_secondary_power', power);
```

The heating power is calculated as follows (openHAB ECMAScript).

```
var power = 0;

var operation_mode = itemRegistry.getItem('espaltherma_operation_mode').state.toString();
var defrost_operation = itemRegistry.getItem('espaltherma_defrost_operaton').state;

if (operation_mode === 'Heating' && defrost_operation === OFF) {
  var flow = parseFloat(itemRegistry.getItem('espaltherma_flow_sensor').state);
  var lw_temp = parseFloat(itemRegistry.getItem('espaltherma_leaving_water_temp_before_buh').state);
  var iw_temp = parseFloat(itemRegistry.getItem('espaltherma_inlet_water_temp').state);

  var diff_temp = lw_temp - iw_temp;
  
  var water_density = 0.993;  // for approx 35°C

  var convert_flow = 60 * water_density; // converts l/min to kg/h for water
  var thermal_capacity = 1.1626; // of water in Wh/(kg•K)

  power = flow * convert_flow * thermal_capacity * diff_temp;
  power = Math.round(power * 100) / 100;
}

events.sendCommand('espaltherma_heating_power', power);
```

These are my graphs of the measured data (Grafana from InfluxDB).

![image](https://user-images.githubusercontent.com/34278179/193429922-90987821-af30-4068-958d-dbdc83555bf0.png)
![image](https://user-images.githubusercontent.com/34278179/193429968-9455bd0f-bd53-4d64-baa4-6a7469a16fcf.png)

What do you think? Would appreciate a discussion about it.

**Comment by sahni2021** [comment]
I also noticed large fluctuations calculating the COP using the primary current, but never thought of any for the secondary current. Now when I use the calculation below (inside Home Assistant) I agree with @dandjo in that it delivers more plausible values.

`{{
((state_attr('sensor.althermasensors','Flow sensor (l/min)')| float * 0.06 * 1.16 * (state_attr('sensor.althermasensors','Outlet Water Heat Exch. Temp. (R1T)') | float - state_attr('sensor.althermasensors','Inlet water temp.(R4T)')|float) )
    /
  (state_attr('sensor.althermasensors','INV secondary current (A)') | float / 3 * state_attr('sensor.althermasensors','Voltage (V)')|float / 1000))
  |round(2)
}}`

**Comment by dandjo** [comment]
What I have seen so far: primary current always precedes the heat generated, i.e. the compressor starts up, produces heat and shortly afterwards this results in higher temperatures. That's not the case with Secondary Current, which makes it implausible again. Also, dividing the current by 3 may not always result in correct values, especially at higher values.

**Comment by sahni2021** [comment]
I think to solve the problem with the current ambiguity the only solution might be to use an external device (e.g. Shelly 3EM) to measure the currents in real-time. Something I rather avoid due the high costs for the Shelly device and the cable clutter in the circuit breaker box ;-)

**Comment by Kepro** [comment]
I think the issue is with `Voltage (N-phase) (V)` it's not stable from 75 it goes up to 320 and my unit is on 1 phase so there should be around 230-240V always... 

<img width="568" alt="image" src="https://user-images.githubusercontent.com/1714370/195584763-e639230d-8df0-4053-b0be-2245cd09c8ad.png">

edit: it looks like with higher temperatures also Voltage is higher causing COP to drop... as I started with COP 12.1 and now it's 2.16 (voltage 340) 

edit2: 57 celsiuz -> voltage 430

**Comment by dandjo** [comment]
Hi @Kepro be aware that N-Phase is the current after the SSR as I understood, not the current of your grid. It's the current after the transformation for the compressor motor itself. It's a PWM signal.

**Comment by Kepro** [comment]
hi @dandjo thanks for the quick reply, I just found a couple of formulas for COP, so I used it and many people complain about the same from my understanding... so is this then correct? https://github.com/raomin/ESPAltherma#calculating-cop

**Comment by dandjo** [comment]
@Kepro compared to my SmartMeter sometimes it's correct, sometimes not. I think the problem is the asynchronous reading respectively the offset the data from the registers. In average it's pretty close.

**Comment by Kepro** [comment]
@dandjo one more question, it's possible to use `Voltage (N-phase) (V)` or `INV secondary current (A)` or `INV primary current (A)` to calculate current load? I have 1 phase external unit 6kW 

also today I'm heating and COP (outside 18C, 33 water temp) is above 10... that seems too high if I'm using formula from readme

**Comment by dandjo** [comment]
`Voltage (N-phase) (V)` is the voltage **after** the "transformer" for the compressor. `INV secondary current (A)` is the current **after** the "transformer" for each phase (there are always 3!). We assume that all 3 phases have the same voltage (I'm not sure if this is the case).

The other option is to mutiply `INV primary current (A)` with the voltage of your grid, since this is the current **before** the "transformer".

I'm not quite sure which method is better or more accurate in the end.

**Comment by pwwukpw** [comment]
Managed to grab a capture with Amps and Volts  showing ..  V way above Grid voltage and both primary and secondary current. Installer did put a grid meter just before the heat pump but I think it's dumb. Any consensus as to which formula to use for accurate COP. With so many variables I'm keen to optimise operation for best lowest consumption. 

![image](https://user-images.githubusercontent.com/88422402/198817830-94d7c9da-163e-49dd-bd51-de3ffa5d40e6.png)

**Comment by raomin** [comment]
I'll move this to the discussion section, and I would welcome any suggestion to improve the COP calculation on the readme :)

**Reply by pwwukpw** [reply]
Don't suppose has an updated COP formula? Especially one that reads from the incoming electrify meter wired upto the monoblock?

**Comment by dg362** [comment]
The discussion seems to have gone a bit stale, but as getting COP is one of the primary reasons I got an M5stickCplus and installed and configured ESPAltherma, I'd like to get to the bottom of this.
I am in the UK and had Octopus Energy install my Daikin ASHP plus Daikin DHW tank (plus buffer and expansion tanks). I'm on a 230V system with ASHP on a 20A fuse and immersion for the DHW tank on a 16A fuse.
I do have the Daikin Madoka simplified heating controller in the lounge as the internal temperature, and there's an outside temperature sensor in the Monobloc. I have the Onecta App and also have HomeAssistant integrated with the base Daikin API via https://github.com/speleolontra/daikin_residential_altherma

I have a Daikin Altherma 3 Monobloc model EDLA06E2V3 using sensors relevant to COP as below. I have other sensors enabled of course.
```
{0x62,9,105,2,-1,"Flow sensor (l/min)"}
{0x61,2,105,2,1,"Leaving water temp. before BUH (R1T)"}
{0x61,8,105,2,1,"Inlet water temp. (R4T)"}
{0x21,0,105,2,-1,"INV primary current (A)"}
{0x21,2,105,2,-1,"INV secondary current (A)"}
{0x21,2,101,2,-1,"Voltage (N-phase) (V)"}
{0x21,4,101,2,-1,"Voltage (V)"}
```

My COP calculation is below (note: I set my MQTT name to differentiate it from the other Daikin API integration).
```
% if is_state_attr('sensor.edla06e2v3_espaltherma','Operation Mode', 'Heating') and is_state_attr('sensor.edla06e2v3_espaltherma','Freeze Protection', 'OFF')  %} 
            {{
            ((state_attr('sensor.edla06e2v3_espaltherma','Flow sensor (l/min)')| float * 0.06 * 1.16 * (state_attr('sensor.edla06e2v3_espaltherma','Leaving water temp. before BUH (R1T)') | float - state_attr('sensor.edla06e2v3_espaltherma','Inlet water temp. (R4T)')|float) )
            /
            (state_attr('sensor.edla06e2v3_espaltherma','INV secondary current (A)') | float * state_attr('sensor.edla06e2v3_espaltherma','Voltage (N-phase) (V)')|float / 1000))
            |round(2) }}
            {% else %} 0.0 {%endif%}
```

For a hot water Comfort schedule from 03:00 to 04:00 with a target tank temperature of 48C and a starting temperature of 44C, the calculated COP at 03:15 was min: 3.22, mean: 3.61, max: 4.09. I've also calculated the COP manually just to see the result and it doesn't feel right for hot water usage.

Example calculation at 03:15 results in COP of **3.29808**
`(25.5*0.06*1.16*(56.9-53.1))/((14.3*143)/1000)`

Changing to `Voltage (V)` instead for the same time period results in a COP of **1.32852**
`(25.5*0.06*1.16*(56.9-53.1))/((14.3*355)/1000)`

Changing to `INV primary current (A)` and then just using 230V as a fixed value for a UK power supply, the result is **2.2556**
`(25.5*0.06*1.16*(56.9-53.1))/((13.0*230)/1000)`

Comparing `INV secondary current (A)` and `Voltage (N-phase) (V)` they are the same value but * 10 for Amps vs. Volts which doesn't seem right. e.g. 135A vs 13.5V. This is why I'm thinking I should be using for the EDLA models the `Voltage (V)` sensor of `0x21,4,101,2,-1`

Comparing `INV secondary current (A)` and `INV primary current (A)` there really isn't much between these although secondary is on average about 1 more than primary.

In neither case do the values match up with the UK standards, accepting the points made earlier that these are not at the source power but after processing within the ASHP so it's not the case that sensors should always be within the UK allowance of 230 V -6%, +10% (range of 216.2 V to 253.0 V).

Is anyone else here from the UK who can share what system they've got plus also their COP calculation? Preferably someone with an EDLA model, and given that is all Octopus Energy is installing presently I'd expect there to be someone out there with a similar setup to me.

**Reply by dandjo** [reply]
As mentioned above, the Voltage is a measurement of the neutral wire after the inverter. It's useless for calculation power. `INV primary current (A)` is the current measurement before the inverter, `INV secondary current (A)` is the current after the inverter. If you want to calculate the power, you have to multiply `INV primary current (A)` with your average grid Voltage.

Since the resolution of the measurement of `INV primary current (A)`  is relatively imprecise (0.5 A accuracy), the calculation is very imprecise, especially for small currents. I endet up installing a Shelly EM with Tasmota, which gives me pretty precise electrical measurements. In openHAB I use those values to calculate the COP.

**Reply by mydliar** [reply]
@dg362 - I also live in the UK and I think I have a very similar setup, EDLA09D(3)V3 installed by Octopus two months ago. I too installed ESPAltherma on M5stickCplus to get more accurate diagnostics.

During the installation I found out that the engineers also installed their own monitoring device using Emoncms (sending the diagnostics data somewhere to Octopus servers presumably). Never heard of it before, but the installers left me the login credentials and I figured out that the device is hooked into both the heat pump as well as the heat pump's input electricity meter, and provides bunch of very useful statistics. It has its own MQTT, which I then bridged to my Home Assistant and suddenly I had all the input electricity and output energy stats at my fingertips:
![image](https://github.com/raomin/ESPAltherma/assets/98688285/12486995-b952-41a3-885f-50a9afa7f675)

I haven't got as far monitoring the COP yet, but my understanding is that I just need to divide the Heatmeter Energy (kWh) by the Electric Meter energy (kWh) and I have the COP (currently 2.75).

Don't you have the same device? It appeared on my LAN as EmonHP. I can also send you a photo what it looks like.

**Reply by dg362** [reply]
I've heard of the Emoncms system, but sadly this doesn't seem to be something they've installed with my setup which was done in July this year. They never asked for my local LAN or Wi-Fi password, and a quick scan of my network doesn't show anything obvious that might be it. If they've installed it for you, then you may also have CT clamps over the individual power connections so you can see the split of power usage per device (ASHP vs Immersion). When I was looking at Emoncms that is partly what I was considering it for.

If you have a photo you could share, I could take a look to see if it is indeed somewhere hidden away, then I could get in touch with them to see about what it's telling me that would supplement what I get from ESPAltherma plus the Diakin native system. Clearly, it's been configured to match specific sensor data, which in itself could be interesting.

For COP, for now, I've set up two entries in my template.yaml file with subtly different names and calculations and will monitor them over time to get a sense of which is giving me something that 'feels right'.

**Reply by stephenacrown** [reply]
@dg362 @dandjo 

Hi guys,

I've written a guide/walkthrough on how I integrated ESPAltherma which is great, with OpenEnergyMonitor EmonCMS cloud service which allows for publishing dashboards publicly (or privately) with full system details including CoP etc.

https://community.openenergymonitor.org/t/daikin-altherma-espaltherma-home-assistant-with-openenergymonitor/25057/6?u=stephen_crown


Here is a "almost real time" link to my Daikin system that data is fed from ESPAltherma:

https://emoncms.org/app/view?name=9ChurchStreet&readkey=53fbf32311dcafcbf15209adc8e10c7f&mode=power&start=1702248840&end=1702451640

This is what it looks like as an example when logged in as the owner:

![image](https://github.com/raomin/ESPAltherma/assets/38353501/22c60c1c-abc8-4183-8796-50137e698373)

Also, We've recently discovered those with a BUH (backup heater) must not use sensor "Leaving Water before BUH R1T" as this reads too high. R2T is the correct sensor (someone has a MID compliant billing grade heat meter to provide correlated data for us). For my system, R2T provides a ~14% lower heat power than using R1T, therefore lower CoP, but is more accurate. 

I have opened an [Issue here](https://github.com/raomin/ESPAltherma/issues/367) relating to this point to see if this can be captured on the main readme as I think the current formula is misleading. 

https://github.com/raomin/ESPAltherma/issues/367

Hope this helps!

cc @raomin

**Comment by mydliar** [comment]
@dg362 - for ref, this is how Emon device looks in my garage:
![IMG20231007180243](https://github.com/raomin/ESPAltherma/assets/98688285/c354c63f-4b72-4ea6-994d-cacea94e52b8)

**Comment by dandjo** [comment]
Here's the comparision of ESPAltherma (green) and my Shelly EM (yellow). At a high load the deviation is minimal and the calculation with values from ESPAltherma is reasonably accurate, but at low power the deviation is relatively high.

![image](https://github.com/raomin/ESPAltherma/assets/34278179/b0d16f02-d2a1-49d6-b967-4d953013fa8e)

The main issue is the Reactive Power, which is quite high, also especially at low power output. So if you calculate the Apparent Power via Voltage * Current, the accuracy is quite low at low power.

![image](https://github.com/raomin/ESPAltherma/assets/34278179/747b29cc-2d99-48fd-b184-f0b8eab2bdfd)

Overall, calculating electrical power with ESPAltherma is an approach that, in my opinion, is not accurate enough for low power heating purposes.

**Comment by johngouk** [comment]
At the risk of being off-topic, I haven't even attempted to calculate power usage from ESPAltherma data, on the basis that it's far simpler to obtain a cheap CT clamp device and use the data from that. I am currently using a [Tuya-connected Dual CT device](https://www.aliexpress.com/item/1005005873738695.html) with which I monitor the main house supply and the HP separately. The main supply monitoring gives me a comparison with the main house utility meter, and it seems pretty close. Admittedly, I had to write some code to do that, but other less complex devices that provide data to HA etc. are available. I should probably write up what I did!

**Comment by abmantis** [comment]
I have an EDLA16DA3W monoblock, which does not provide the "Voltage (Phase N)" attribute. I was using "INV Primary Current" multiplied by 230 * 3 (its a 3 phase installation), but I was getting values a quite a bit higher that the ones that the heatpump HMI shows.
I'm now testing using 230 * "INV Secondary Current", and the values are very close to the ones that show on the HMI!

**Reply by LilSlippinJimmy** [reply]
I've found the same thing. Secondary Current is much closer to what the Heat Pump actual usage is

**Reply by THK56** [reply]
Comment to the formula: current * voltage * 3 isn't correct. In case of real sine waves for current and voltage and a power factor of 100%, the formula would be: current * voltage * square root of 3 (= 1.71...). Both power-factor and wave-form are not really known. For the wave-form we can assume a more or less correct sine wave. The power-factor is about 80 - 90% (in my case). As an approach you can take 85%. Therefor as an approximation you can use: INV current * INV voltage * 0.85 * 1.71 (knowing that all the other power consumption for controlling the hp is not considered in the COP calculation). An external power meter is still a more accurate option.

**Reply by dandjo** [reply]
Secondary Current seems to fit better since your calculation of a three phase system with one current value is definitely wrong, as THK56 said. It also depends how the inverter shifts voltage and current for the compressor motor. I guess the compressor for your monoblock is a different model than for my split machine.

**Comment by crag364** [comment]
My unit does not provide Voltage (Phase N), and it is only a single phase unit. What is the best way to calculate COP? Currently I am using this but am not sure how accuate it is. 

- name: "COP"
      unique_id: "espaltherma_cop"
      unit_of_measurement: 'COP'
      state: "{% if is_state_attr('sensor.altherma','Operation Mode', 'Heating') and is_state_attr('sensor.altherma','Freeze Protection', 'OFF')  %} 
{{ 
  ((state_attr('sensor.altherma','Flow sensor (l/min)')| float * 0.06 * 1.16 * (state_attr('sensor.altherma','Leaving water temp. before BUH (R1T)') | float - state_attr('sensor.altherma','Inlet water temp.(R4T)')|float) )
    /
  (state_attr('sensor.altherma','INV primary current (A)') | float * state_attr('sensor.altherma','INV secondary current (A)')|float * 230 / 1000))
  |round(2)
}}
{% else %} 0 {%endif%}"

**Comment by jsonoine1111** [comment]
First of all I would like to thank for fantastic integration and ideas in this forum. 
I am happy to add my input  to the subject so maybe it will help other users to get expected results quicker.
 
For power measurement I purchased small energy meter "Zemismart 3 Phase 4 Wire Zigbee" and using it in HA ZHA (Price ~40EUR on Ali).
In my setup I use 3 Current Circuit Breakers, 1 diff ABB SH204 first and 2 ABB SH203 after it to Altherma 3 (3phase 16kW with 9kw BUH). One SH203 Current Circuit Breaker for heat pump, another to Backup Heater BUH 9kW (configured as 3kw stage 1+3kw stage 2 for now, need to improve wiring) and diff controls total leakages (as unit is located to outdoor and is related to water). 
Zemismart 3 Phase 4 Wire Zigbee is having Total Power metrics which can be used directly in calculation for Energy Input metric (currently in equation calculated as =current * voltage / 3 / 1000). But there are also voltage , current and power for each phase available.

For honest COP calculation I believe all input consuming energy should be calculated including defrost (with minus sign) and DWH, because defrost operation is part of the consumed energy and the bill. Separate COP metrics can be created for Heating, DWH and Service (defrost, emergency etc) if required.

PS Previously before sensor installation 'INV secondary current (A)' metric was closer to real valu in my setup.

![Zemismart_energy_meter_ZHA](https://github.com/user-attachments/assets/d3845817-2e86-467e-b48a-6bb6291912e5)

![altherma3_energy_meter_zha_lq](https://github.com/user-attachments/assets/7eced84a-9a1e-4a9b-90a5-89aa435cda08)
